File: src/main/java/no/skatteetaten/aurora/databasehotel/dao/oracle/OracleDatabaseManager.java
package no.skatteetaten.aurora.databasehotel.dao.oracle;

import static java.lang.String.format;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import javax.sql.DataSource;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import no.skatteetaten.aurora.databasehotel.dao.DataAccessException;
import no.skatteetaten.aurora.databasehotel.dao.DatabaseManager;
import no.skatteetaten.aurora.databasehotel.dao.DatabaseSupport;
import no.skatteetaten.aurora.databasehotel.dao.dto.Schema;

public class OracleDatabaseManager extends DatabaseSupport implements DatabaseManager {

    private static Logger logger = LoggerFactory.getLogger(OracleDatabaseManager.class);

    public OracleDatabaseManager(DataSource dataSource) {

        super(dataSource);
    }

    private static String convertToValid(String schemaName) {

        return schemaName.toUpperCase();
    }

    @Override
    public boolean schemaExists(String schemaName) {

        String query = "SELECT * FROM dba_users u WHERE username=?";
        return getJdbcTemplate().queryForList(query, schemaName).size() == 1;
    }

    @Override
    public String createSchema(String schemaName, String password) {

        String schemaNameValid = convertToValid(schemaName);
        String dataFolder = getDataFolder();
        String[] createSchemaStatements = {
            format("create bigfile tablespace %s datafile '%s/%s.dbf' size 10M autoextend on maxsize 1000G",
                schemaNameValid, dataFolder, schemaNameValid),
            format("create user %s identified by %s default tablespace %s", schemaNameValid, password, schemaNameValid),
            format("grant connect,resource to %s", schemaNameValid),
            format("grant create view to %s", schemaNameValid),
            format("alter user %s quota unlimited on %s", schemaNameValid, schemaNameValid),
            format("alter user %s profile APP_USER", schemaNameValid)
        };
        executeStatements(createSchemaStatements);

        return schemaNameValid;
    }

    /**
     * When changing the password we will also make sure that the account is open (in case it had become locked during
     * a transition for the previous password from active processes).
     *
     * @param schemaName the schema to update the password for
     * @param password   the new password
     */
    @Override
    public void updatePassword(String schemaName, String password) {

        String[] statements = {
            format("ALTER USER %s IDENTIFIED BY %s", schemaName, password),
            format("alter user %s account unlock", schemaName)
        };
        executeStatements(statements);
    }

    @Override
    public Optional<Schema> findSchemaByName(String schemaName) {

        String query = "SELECT * FROM dba_users u WHERE username=?";
        return queryForOne(query, Schema.class, schemaName);
    }

    @Override
    public List<Schema> findAllNonSystemSchemas() {

        String currentUserName = getJdbcTemplate().queryForObject("select user from dual", String.class);
        String query = "SELECT * FROM dba_users u WHERE default_tablespace not in ('SYSTEM', 'SYSAUX', 'USERS') "
            + "and default_tablespace=username and username!=?";
        return queryForMany(query, Schema.class, currentUserName);
    }

    @Override
    public void deleteSchema(String schemaName) {

        disconnectAllUsers(schemaName);

        String[] dropSchemaStatements = {
            format("DROP USER %s cascade", schemaName),
            format("DROP TABLESPACE %s INCLUDING CONTENTS AND DATAFILES", schemaName)
        };
        executeStatementsOnlyLogErrors(dropSchemaStatements);
    }

    /**
     * Will try to disconnect all active users and connections against a specified schema.
     * <p>
     * Disconnecting all users for a schema is actually more flaky and unreliable than it ideally should be, so this
     * method will try a few times before giving up.
     *
     * @param schemaName the name of the schema to disconnect users from
     * @throws DataAccessException if all sessions could not be closed by the nth attempt.
     */
    private void disconnectAllUsers(String schemaName) throws DataAccessException {

        int maxNumberOfDisconnectAttempts = 10;
        // We may not be able to disconnect all users on our first try, so try a few times before moving on.
        for (int attempt = 1; attempt <= maxNumberOfDisconnectAttempts; attempt++) {
            List<Map<String, Object>> activeSessions =
                getJdbcTemplate().queryForList("SELECT s.sid, s.serial# FROM v$session s where username=?", schemaName);
            if (activeSessions.isEmpty()) {
                // Only when there are no active sessions are we actually successfully finished.
                return;
            }
            activeSessions.forEach(session -> {
                Object sid = session.get("sid");
                Object serial = session.get("serial#");
                String[] statements = {
                    format("ALTER SYSTEM KILL SESSION '%s, %s' IMMEDIATE", sid, serial),
                    // KILL SESSION by it self does not always do the trick...
                    format("ALTER SYSTEM DISCONNECT SESSION '%s, %s' IMMEDIATE", sid, serial)
                };
                executeStatementsOnlyLogErrors(statements);
            });
            if (attempt >= 2) {
                // If we are beyond the first attempt, lets wait a little to let the database server catch its breath.
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                }
            }
        }
        throw new DataAccessException(format("Unable to disconnect all users from schema=%s by "
            + "the %d attempt. Gave up.", schemaName, maxNumberOfDisconnectAttempts));
    }

    private String getDataFolder() {

        String dataFolderQuery =
            "SELECT SUBSTR(FILE_NAME, 1, INSTR(FILE_NAME, '/', -1) -1) as DATA_FOLDER FROM DBA_DATA_FILES where "
                + "TABLESPACE_NAME='SYSTEM'";
        return getJdbcTemplate().queryForObject(dataFolderQuery, String.class);
    }

    private void executeStatementsOnlyLogErrors(String... statements) {

        for (String statement : statements) {
            try {
                getJdbcTemplate().execute(statement);
            } catch (Exception e) {
                logger.warn(format("Error executing statement=[%s]", statement), e);
            }
        }
    }
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/dao/oracle/OracleDatabaseManager.kt
package no.skatteetaten.aurora.databasehotel.dao.oracle

import java.sql.SQLException
import javax.sql.DataSource
import no.skatteetaten.aurora.databasehotel.dao.DataAccessException
import no.skatteetaten.aurora.databasehotel.dao.DatabaseManager
import no.skatteetaten.aurora.databasehotel.dao.DatabaseSupport
import no.skatteetaten.aurora.databasehotel.dao.Schema
import no.skatteetaten.aurora.databasehotel.dao.toSchema
import org.apache.commons.lang3.exception.ExceptionUtils
import org.springframework.dao.EmptyResultDataAccessException

class OracleDatabaseManager(dataSource: DataSource) : DatabaseSupport(dataSource), DatabaseManager {

    private val dataFolder: String?
        get() {

            val dataFolderQuery =
                "SELECT SUBSTR(FILE_NAME, 1, INSTR(FILE_NAME, '/', -1) -1) as DATA_FOLDER FROM DBA_DATA_FILES where " + "TABLESPACE_NAME='SYSTEM' " + "and rownum=1"
            return jdbcTemplate.queryForObject(dataFolderQuery, String::class.java)
        }

    override fun schemaExists(schemaName: String): Boolean {

        val query = "SELECT * FROM dba_users u WHERE username=?"
        return jdbcTemplate.queryForList(query, schemaName).size == 1
    }

    override fun createSchema(schemaName: String, password: String): String {

        val schemaNameValid = convertToValid(schemaName)
        val dataFolder = dataFolder
        val createSchemaStatements = arrayOf(
            "create bigfile tablespace $schemaNameValid datafile '$dataFolder/$schemaNameValid.dbf' size 10M autoextend on maxsize 1000G",
            "create user $schemaNameValid identified by $password default tablespace $schemaNameValid",
            "grant connect,resource to $schemaNameValid",
            "grant create view to $schemaNameValid",
            "alter user $schemaNameValid quota unlimited on $schemaNameValid",
            "alter user $schemaNameValid profile APP_USER"
        )

        try {
            executeStatements(*createSchemaStatements)
            return schemaNameValid
        } catch (e: org.springframework.dao.DataAccessException) {
            throw DataAccessException(e.getDbMessage(), e)
        }
    }

    /**
     * When changing the password we will also make sure that the account is open (in case it had become locked during
     * a transition for the previous password from active processes).
     *
     * @param schemaName the schema to update the password for
     * @param password the new password
     */
    override fun updatePassword(schemaName: String, password: String) {

        val statements = arrayOf(
            "ALTER USER $schemaName IDENTIFIED BY $password",
            "alter user $schemaName account unlock"
        )
        executeStatements(*statements)
    }

    override fun findSchemaByName(schemaName: String): Schema? {
        val query = "SELECT username, last_login as lastLogin FROM dba_users u WHERE username=?"
        return try {
            jdbcTemplate.queryForObject(query, toSchema, schemaName)
        } catch (e: EmptyResultDataAccessException) {
            null
        }
    }

    override fun findAllNonSystemSchemas(): List<Schema> {

        val currentUserName = jdbcTemplate.queryForObject("select user from dual", String::class.java)
        val query = (
            """SELECT username, last_login as lastLogin FROM dba_users u WHERE 
            default_tablespace not in ('SYSTEM', 'SYSAUX', 'USERS', 'MAPTEST', 'AOS_API_USER', 'RESIDENTS') 
            and default_tablespace=username and username!=?""".trimIndent())
        return jdbcTemplate.query(query, toSchema, currentUserName)
    }

    override fun deleteSchema(schemaName: String) {

        // NOOP
    }

    override fun getMaxTablespaces(): Int? {
        val query = "SELECT VALUE FROM v\$parameter WHERE NAME='db_files'"
        return jdbcTemplate.queryForObject(query, Int::class.java)
    }

    override fun getUsedTablespaces(): Int? {
        val query = "SELECT count(*) FROM v\$tablespace"
        return jdbcTemplate.queryForObject(query, Int::class.java)
    }

    companion object {

        private fun convertToValid(schemaName: String): String {
            return schemaName.toUpperCase()
        }
    }
}

fun Throwable.getDbMessage(): String {
    val dbMessage = ExceptionUtils.getThrowableList(this)
        .filterIsInstance<SQLException>()
        .map { it.message }
        .firstOrNull() ?: ""

    return "${this.message} $dbMessage"
}
